### PR TITLE
Adds decorators for Python References and extends Python Reference testing

### PR DESCRIPTION
**Pull request by mruberry**
This PR does the following...

Tests:
- fixes test_type_promotion in test_binary_ufuncs to correctly generate scalar cpu tensors
- fixes test_python_reference_consistency to use the Python Reference's reference inputs
- extends Python reference testing to test_conj_view, test_neg_view, and test_neg_conj_view
- adds a NaN propagation sample input for elementwise unary and binary operations
- fixes the UnaryUfuncInfo class to properly register its reference inputs
- Updates the Python Reference OpInfos to skip error inputs when their behavior on scalar inputs is inconsistent with their reference operators

Code organization:
- moves elementwise type promotion functionality to prims.utils

Prims & Refs:
- fixes scalar cpu tensor handling by having them pass through broadcasting and device and shape checks
- adds two decorators, `elementwise_type_promotion_wrapper` and `out_wrapper`, the former allows for elementwise type promotion to be automated and the latter automatically adds the out kwarg and handles it properly

cc @ezyang who also had some thoughts on cpu scalar tensor handling
cc @chillee -- might want to use this new decorator as we converge decompositions and references

### CODE REVIEW

**Review by ngimel — APPROVED** [review]

File: torch/_refs/__init__.py [hunk 1]
```diff
@@ -811,140 +480,114 @@ def float_power(
...
-    result = prims.sub(a, b)
-
-    (result,) = _convert_dtype(result, dtype=result_dtype)
-
-    if out is not None:
-        out = _maybe_resize_out(out, result.shape)
-        return copy_to(out, result, allow_cross_device=False)  # type: ignore[arg-type]
+        dtype = a.dtype if isinstance(a, TensorLike) else b.dtype  # type: ignore[union-attr]
+        python_type = utils.dtype_to_type(dtype)
+        if not utils.is_weakly_lesser_type(type(alpha), python_type):
```
Comment (ngimel) [inline torch/_refs/__init__.py:575]: why do we need both `is_weakly_lesser_type` and `can_safely_cast_to` that do the same thing?

Comment (mruberry) [reply torch/_refs/__init__.py:575]: There's probably an opportunity to unify them but one currently works on types and the other on dtypes.

Comment (ngimel) [reply torch/_refs/__init__.py:575]: Right, but you have maps to convert between the two.

**Pull request closed without merging**